internal/statushistory/statushistory.go

```diff
@@ -75,11 +81,163 @@ func (s *StatusHistory) RecordStatus(ctx context.Context, ns Namespace, status s
 		Data:    status.Data,
 	})
 }
+
+// HistoryRecord represents a single record of status information.
+type HistoryRecord struct {
+	ModelUUID model.UUID
+	Kind      status.HistoryKind
+	Tag       string
+	Status    status.DetailedStatus
+}
+
+// Scanner is an interface for reading lines from a source.
+type Scanner interface {
+	LineBytes() (line []byte, pos int, err error)
+	Close() error
+}
+
+// StatusHistoryReader is a reader for status history records.
+// It reads records from an io.Reader and unmarshals them into Record structs.
+type StatusHistoryReader struct {
+	modelUUID model.UUID
+	scanner   Scanner
+}
+
+// NewStatusHistoryReader creates a new StatusHistoryReader that reads from the
+// given io.Reader.
+func NewStatusHistoryReader(modelUUID model.UUID, scanner Scanner) *StatusHistoryReader {
+	return &StatusHistoryReader{
+		modelUUID: modelUUID,
+		scanner:   scanner,
+	}
+}
+
+// StatusHistoryReaderFromFile creates a new StatusHistoryReader that reads from
+// the given file path. It opens the file for reading and returns a
+// StatusHistoryReader.
+func ModelStatusHistoryReaderFromFile(modelUUID model.UUID, path string) (*StatusHistoryReader, error) {
+	file, err := os.OpenFile(path, os.O_RDONLY, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	size, err := file.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewStatusHistoryReader(modelUUID, scannerCloser{
+		Scanner: backscanner.New(file, int(size.Size())),
+		Closer:  file,
+	}), nil
+}
+
+type jsonRecord struct {
+	ModelUUID model.UUID        `json:"model-uuid"`
+	Labels    map[string]string `json:"labels"`
+}
+
+// Walk reads the status history records from the reader and applies the
+// given function to each record.
+func (r *StatusHistoryReader) Walk(fn func(HistoryRecord) (bool, error)) error {
+	kinds := status.AllHistoryKind()
+
+	// Read each line of the log file and unmarshal it into a LogRecord.
+	// Filter out records that do not match the requested entities.
+	for {
+		line, _, err := r.scanner.LineBytes()
+		if errors.Is(err, io.EOF) {
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		var rec jsonRecord
+		if err := json.Unmarshal(line, &rec); err != nil {
+			continue
+		}
+
+		// Check if the record belongs to the current model.
+		if rec.ModelUUID == "" || rec.ModelUUID != r.modelUUID {
+			continue
+		}
+
+		// If the record does not have the requested labels, skip it.
+		if len(rec.Labels) == 0 || rec.Labels[categoryKey] != statusHistoryCategory {
+			continue
+		}
+
+		kind := status.HistoryKind(rec.Labels[kindKey])
+		if _, valid := kinds[kind]; !valid {
+			continue
+		}
+
+		var data map[string]any
+		if labelData := rec.Labels[dataKey]; len(labelData) > 0 {
+			if err := json.Unmarshal([]byte(labelData), &data); err != nil {
+				continue
+			}
+		}
+
+		var since time.Time
+		if sinceStr := rec.Labels[sinceKey]; len(sinceStr) > 0 {
+			if t, err := time.Parse(time.RFC3339, sinceStr); err == nil {
+				since = t
```

Comment:
Why are we ok with time missing but anything else results in an ignored "invalid" line?

Reply:
Time and data is optional in status. Everything else should be there. Invalid data should result in an error though.